analysis/cross_partition_combiners.py

```diff
@@ -94,3 +93,56 @@ def _sum_metrics_to_metric_utility(
         ratio_data_dropped=data_dropped,
         absolute_error=absolute_error,
         relative_error=relative_error)
+
+
+def _partition_selection_per_to_cross_partition(
+        prob_keep: float) -> metrics.PrivatePartitionSelectionUtility:
+    return metrics.PrivatePartitionSelectionUtility(
+        strategy=None,
+        num_partitions=1,
+        dropped_partitions=metrics(mean=prob_keep,
+                                   var=prob_keep * (1 - prob_keep)),
+        ratio_dropped_data=0)  # todo: implement
+
+
+def _add_dataclasses_by_fields(dataclass1, dataclass2,
+                               fields_to_ignore: List[str]) -> None:
+    """Recursively adds all numerical fields of one dataclass to another.
+
+    The result is in dataclass1.
+
+    Assumptions:
+      1. dataclass1 and dataclass2 are instances of the same dataclass type.
+      2. all fields which should be processed (i.e. not in fields_to_ignore)
+       are dataclasses or '+' operator can be applied.
+      3. For all dataclasses fields assumptions 1,2 apply.
+
+    Attributes:
+        dataclass1, dataclass2: instances of the same dataclass type.
+        fields_to_ignore: field names which should be ignored.
+    """
+    assert type(dataclass1) == type(dataclass2)
+    fields = dataclasses.fields(dataclass1)
+    for field in fields:
+        if field.name in fields_to_ignore:
+            continue
+        value1 = getattr(dataclass1, field.name)
+        value2 = getattr(dataclass2, field.name)
+        if dataclasses.is_dataclass(value1):
+            _add_dataclasses_by_fields(value1, value2, fields_to_ignore)
+            continue
+        setattr(dataclass1, field.name, value1 + value2)
+
+
+def _multiply_float_dataclasses_field(dataclass, factor: float):
+    """Recursively multiply all float fields of the dataclass by given number."""
+    fields = dataclasses.fields(dataclass)
+    for field in fields:
+        value = getattr(dataclass, field.name)
+        if value is None:
+            continue
+        if field.type is float:
+            setattr(dataclass, field.name, value * factor)
+        if dataclasses.is_dataclass(value):
+            _multiply_float_dataclasses_field(value, factor)
+            continue
```

Comment:
Optional:
```python
if value is not None and field.type is float:
    setattr(dataclass, field.name, value * factor)
if dataclasses.is_dataclass(value):
    _multiply_float_dataclasses_field(value, factor)
```

I don't think you need the continue at the end.

Also, what happens if you do `None * factor`? Maybe it is handled gracefully and results in `None`?

Reply:
Thx, we don't need, it was left after some refactroing ))

`None * factor` throws errors